pkg/v1/tkg/tkgconfigupdater/save.go

```diff
@@ -11,91 +11,117 @@ import (
-func SaveConfig(clusterConfigPath string, rw tkgconfigreaderwriter.TKGConfigReaderWriter, configObj interface{}) error {
+func encodeMap(data map[string]string) {
+	for key, value := range data {
+		if shouldEncode(key) {
+			encodedValue, err := encodeValueIfRequired(value)
+			if err == nil {
+				data[key] = encodedValue
+			}
+		}
+	}
+}
+
+func shouldEncode(key string) bool {
+	return utils.ContainsString(KeysToEncode, key)
+}
+
+// SaveConfig saves configuration to ReaderWriter and/or to File
+// allows clusterConfigPath to be empty, in which case no file is written
+// allows TKGConfigReaderWriter to be empty, in which case no update is made to that writer
+// (allows both to be empty, in which case the method does nothing)
+func SaveConfig(clusterConfigPath string, tkgConfigReaderWriter tkgconfigreaderwriter.TKGConfigReaderWriter, configObj interface{}) error {
+	configMap, err := CreateConfigMap(configObj)
+	if err != nil {
+		return err
+	}
+
+	if tkgConfigReaderWriter != nil {
+		tkgConfigReaderWriter.SetMap(configMap)
+	}
+	if len(clusterConfigPath) > 0 {
+		err = SetConfig(configMap, clusterConfigPath)
+	}
+
+	return err
+}
+
+// CreateConfigMap transforms a configuration object into a map
+func CreateConfigMap(configObj interface{}) (map[string]string, error) {
+	// turn the object into a YAML byte array
 	configByte, err := yaml.Marshal(configObj)
 	if err != nil {
-		return errors.Wrap(err, "unable to marshal config object")
+		return nil, errors.Wrap(err, "unable to marshal config object")
 	}
 
+	// turn the byte array into a map
 	configMap := make(map[string]string)
 
 	err = yaml.Unmarshal(configByte, &configMap)
 	if err != nil {
-		return errors.Wrap(err, "error unmarshalling file")
+		return nil, errors.Wrap(err, "error unmarshalling file")
 	}
 
-	for k, v := range configMap {
-		if utils.ContainsString(KeysToEncode, k) {
-			encodedValue, err := encodeValueIfRequired(v)
-			if err == nil {
-				v = encodedValue
-			}
-		}
-		err = os.Setenv(k, v)
-		if err != nil {
-			log.Warningf("%s is not set to environment variable properly", k)
-		}
-		// sets variable in viper store
-		rw.Set(k, v)
+	// certain values (like passwords) need encoding
+	encodeMap(configMap)
 
-		if utils.ContainsString(KeysToNeverPersist, k) {
-			continue
-		}
-		// sets variable in config file
-		err = SetVariableInConfig(k, v, "", clusterConfigPath)
-		if err != nil {
-			return err
-		}
-	}
-
-	return nil
+	return configMap, nil
 }
 
-// SetVariableInConfig allows to explicit override of config value
-// with updating the value in provided tkg-config file
-// i.e. it will add the "key: value" in the tkg-config file if key is not present
-// if key is present it will update the value of the key in the file
-func SetVariableInConfig(key, value, comment, clusterConfigPath string) error {
+// SetConfig saves map of key-value pairs to config file (creating if nec)
+func SetConfig(data map[string]string, clusterConfigPath string) error {
```

Comment:
I feel this is a very critical function overall. And if possible we should try to add few unit tests around this if does not exist already.